packages/editor-ui/src/components/InlineExpressionEditor/InlineExpressionTip.vue

```diff
@@ -1,127 +1,185 @@
-		</n8n-text>
-		<n8n-text size="small" :class="$style.text"
-			>{{ $locale.baseText('expressionTip.noExecutionData') }}
-		</n8n-text>
-	</div>
-
-	<div v-else :class="$style.tip">
-		<n8n-text size="small" :class="$style.tipText"
-			>{{ $locale.baseText('parameterInput.tip') }}:
-		</n8n-text>
-		<n8n-text size="small" :class="$style.text">
-			{{ i18n.baseText('parameterInput.anythingInside') }}
-		</n8n-text>
-		<code v-text="`{{ }}`"></code>
-		<n8n-text size="small" :class="$style.text">
-			{{ i18n.baseText('parameterInput.isJavaScript') }}
-		</n8n-text>
-		<n8n-link
-			:class="$style['learn-more']"
-			size="small"
-			underline
-			theme="text"
-			:to="expressionsDocsUrl"
-		>
-			{{ i18n.baseText('parameterInput.learnMore') }}
-		</n8n-link>
-	</div>
-</template>
-
 <script setup lang="ts">
 import { useI18n } from '@/composables/useI18n';
 import { useNDVStore } from '@/stores/ndv.store';
-import { computed } from 'vue';
-import { EXPRESSIONS_DOCS_URL } from '@/constants';
+import { computed, ref, watch } from 'vue';
+import { EditorSelection, EditorState, type SelectionRange } from '@codemirror/state';
+import { type Completion, CompletionContext } from '@codemirror/autocomplete';
+import { datatypeCompletions } from '@/plugins/codemirror/completions/datatype.completions';
+import { watchDebounced } from '@vueuse/core';
+import { FIELDS_SECTION } from '@/plugins/codemirror/completions/constants';
+import { isCompletionSection } from '@/plugins/codemirror/completions/utils';
+
+type TipId = 'executePrevious' | 'drag' | 'default' | 'dotObject' | 'dotPrimitive';
+
+type Props = {
+	editorState?: EditorState;
+	unresolvedExpression?: string;
+	selection?: SelectionRange;
+};
+
+const props = withDefaults(defineProps<Props>(), {
+	editorState: undefined,
+	unresolvedExpression: '',
+	selection: () => EditorSelection.cursor(0),
+});
 
 const i18n = useI18n();
 const ndvStore = useNDVStore();
 
-const props = defineProps<{ tip?: 'drag' | 'default' }>();
+const canAddDotToExpression = ref(false);
+const resolvedExpressionHasFields = ref(false);
+
+const canDragToFocusedInput = computed(
+	() => !ndvStore.isDNVDataEmpty('input') && ndvStore.focusedMappableInput,
+);
 
-const tip = computed(() => {
-	if (!ndvStore.hasInputData) {
+const emptyExpression = computed(() => props.unresolvedExpression.trim().length === 0);
+
+const tip = computed<TipId>(() => {
+	if (!ndvStore.hasInputData && ndvStore.isInputParentOfActiveNode) {
 		return 'executePrevious';
 	}
 
-	if (props.tip) return props.tip;
+	if (canAddDotToExpression.value) {
+		return resolvedExpressionHasFields.value ? 'dotObject' : 'dotPrimitive';
+	}
 
-	if (ndvStore.focusedMappableInput) return 'drag';
+	if (canDragToFocusedInput.value && emptyExpression.value) return 'drag';
 
 	return 'default';
 });
-const expressionsDocsUrl = EXPRESSIONS_DOCS_URL;
+
+const cursor = computed(() => props.selection.anchor);
+
+function getCompletionsWithDot(): readonly Completion[] {
```

Comment:
An index calculation might be wrong somewhere - if you type an expression and delete it, these show up in the console:

![Capture 2024-03-22 at 14 32 25@2x](https://github.com/n8n-io/n8n/assets/44588767/0768a9a7-88e6-4abc-84d1-6c59449d52f7)

Reply:
Added some more safety for this case 😄